File: OneToOneCallSample/app/src/main/java/com/stringee/apptoappcallsample/stringee/manager/CallStateMachine.java
package com.stringee.apptoappcallsample.stringee.manager;

import com.stringee.apptoappcallsample.stringee.common.CallStatus;

/** Combines signaling and media state so STARTED is emitted only when both are connected. */
final class CallStateMachine {
    private CallStatus status;
    private boolean answered;
    private boolean mediaConnected;
    private boolean startedOnce;

    CallStateMachine(CallStatus initialStatus) {
        status = initialStatus;
    }

    CallStatus onSignalingAnswered() {
        answered = true;
        return updateConnectedState();
    }

    CallStatus onMediaConnected() {
        mediaConnected = true;
        return updateConnectedState();
    }

    CallStatus onMediaDisconnected() {
        mediaConnected = false;
        if (startedOnce && answered) {
            status = CallStatus.RECONNECTING;
        }
        return status;
    }

    CallStatus setStatus(CallStatus newStatus) {
        status = newStatus;
        return status;
    }

    CallStatus getStatus() {
        return status;
    }

    private CallStatus updateConnectedState() {
        if (answered && mediaConnected) {
            startedOnce = true;
            status = CallStatus.STARTED;
        } else {
            status = CallStatus.STARTING;
        }
        return status;
    }
}

File: Kotlin-OneToOneCallSample/app/src/main/java/com/stringee/kotlin_onetoonecallsample/stringee/manager/CallStateMachine.kt
package com.stringee.kotlin_onetoonecallsample.stringee.manager

import com.stringee.kotlin_onetoonecallsample.stringee.common.CallStatus

/** Combines signaling and media so STARTED is emitted only when both are connected. */
internal class CallStateMachine(var status: CallStatus) {
    private var answered = false
    private var mediaConnected = false
    private var startedOnce = false

    fun onSignalingAnswered(): CallStatus {
        answered = true
        return updateConnectedState()
    }

    fun onMediaConnected(): CallStatus {
        mediaConnected = true
        return updateConnectedState()
    }

    fun onMediaDisconnected(): CallStatus {
        mediaConnected = false
        if (startedOnce && answered) {
            status = CallStatus.RECONNECTING
        }
        return status
    }

    fun setStatus(newStatus: CallStatus): CallStatus {
        status = newStatus
        return status
    }

    private fun updateConnectedState(): CallStatus {
        if (answered && mediaConnected) {
            startedOnce = true
            status = CallStatus.STARTED
        } else {
            status = CallStatus.STARTING
        }
        return status
    }
}
